File: ferias_app/lib/models/factura_detalle.dart
import 'model_parsers.dart';
import 'producto.dart';

class FacturaDetalle {
  final int id;
  final int facturaId;
  final int productoId;
  final String descripcionProducto;
  final double cantidad;
  final double precioUnitario;
  final double subtotalLinea;
  final Producto? producto;

  const FacturaDetalle({
    required this.id,
    required this.facturaId,
    required this.productoId,
    required this.descripcionProducto,
    required this.cantidad,
    required this.precioUnitario,
    required this.subtotalLinea,
    this.producto,
  });

  factory FacturaDetalle.fromJson(Map<String, dynamic> json) {
    return FacturaDetalle(
      id: parseInt(json['id']),
      facturaId: parseInt(json['factura_id']),
      productoId: parseInt(json['producto_id']),
      descripcionProducto: parseString(json['descripcion_producto']) ?? '',
      cantidad: parseDouble(json['cantidad']),
      precioUnitario: parseDouble(json['precio_unitario']),
      subtotalLinea: parseDouble(json['subtotal_linea']),
      producto: json['producto'] is Map<String, dynamic>
          ? Producto.fromJson(json['producto'] as Map<String, dynamic>)
          : json['producto'] is Map
          ? Producto.fromJson(Map<String, dynamic>.from(json['producto'] as Map))
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return <String, dynamic>{
      'id': id,
      'factura_id': facturaId,
      'producto_id': productoId,
      'descripcion_producto': descripcionProducto,
      'cantidad': cantidad,
      'precio_unitario': precioUnitario,
      'subtotal_linea': subtotalLinea,
      'producto': producto?.toJson(),
    };
  }
}

File: app/Models/FacturaDetalle.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class FacturaDetalle extends Model
{
    use HasFactory;

    /** @var list<string> */
    protected $fillable = [
        'factura_id',
        'producto_id',
        'descripcion_producto',
        'cantidad',
        'precio_unitario',
        'subtotal_linea',
    ];

    /**
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'cantidad' => 'decimal:1',
            'precio_unitario' => 'decimal:2',
            'subtotal_linea' => 'decimal:2',
        ];
    }

    public function factura(): BelongsTo
    {
        return $this->belongsTo(Factura::class);
    }

    public function producto(): BelongsTo
    {
        return $this->belongsTo(Producto::class);
    }
}
